controller/src/main/java/io/pravega/controller/store/client/StoreClientFactory.java

```diff
@@ -115,7 +115,7 @@ public ZooKeeper newZooKeeper(String connectString, int sessionTimeout, Watcher
                 this.client = new ZooKeeper(connectString, sessionTimeout, watcher, canBeReadOnly);
                 return this.client;
             } else {
-                Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
+                log.info("Former ZK connection string: {}. New ZK connection string {}.", this.connectString, connectString);
```

Comment:
If the connect string differs, then I believe the underlying zk client won't have that connect string (we don't recreate the underlying client). As such, I have two questions:

- Is there any condition under which the connect string actually changes? I'm not talking about giving IP addresses rather than hostnames like described in this PR, but instead really changing the list of servers. A zk ensemble can change its composition with dynamic configuration, but I don't know whether curator will be propagating it through this API.

Reply:
in this case, say curator propagates new zk ensemble via this api, then we can simply call `client.updateZkServers` with new connection string. 

However, from what i have experimented with, the `newZookeeper` on factory is called only during session "lost" from curator as it attempts to create a new client with a new zksession. 
Since we do not wish to create a new session, any other parameter being sent here is of no interest to us irrespective of whether it is with same or different connection string. so we could ignore the connection string altogether.

Reply:
Here is the `LOST` documentation:

https://curator.apache.org/apidocs/org/apache/curator/framework/state/ConnectionState.html#LOST

and `newZooKeeper` is expected to create a new zk client instance:

https://curator.apache.org/apidocs/org/apache/curator/utils/ZookeeperFactory.html#newZooKeeper-java.lang.String-int-org.apache.zookeeper.Watcher-boolean-

Now, if this is only invoked when the session has expired, then executing this method is pointless, and I'm concerned that we will be leaving the controller instance in a bad state as it will keep trying to get a valid zk client instance and getting a bad zk client instance instead.

 Why don't we simply abort this controller instance in the case it asks for a new zookeeper instance and let the controller instance restart?

Reply:
Thats exactly what happens - as we are restarting controller, any attempt from curator to silently create a new zookeeper client and hence a session is thwarted by this little trick of having our own factory.

Note: as shutdown is happening, different modules continue their work until it is all cancelled. And if they get a handle to a working zookeeper client during this time, they may be able to perform store updates despite a session expiration.. hence we need to prevent it.

Reply:
According to the behavior observed in the longevity run and a discussion with @shiveshr:

If the old `connectString` is invalid after it has changed, then the current Curator client will eventually set the connection state to `LOST` (the ZK ensemble will be unreachable) and this will induce a graceful restart of the Controller. With the restart, a new client will be created with the new `connectString`. This is the current behavior of the Controller with this PR as observed in the longevity run, which I believe that is the expected one. Moreover, I feel this approach more natural than forcing an instance abort directly in `newZookeeper`.

Reply:
From what you are saying, if I replace the implementation of that `newZooKeeper` method with:

```
public ZooKeeper newZooKeeper(String connectString, int sessionTimeout, Watcher watcher, boolean canBeReadOnly) {
   return null
}
```

I have the precise behavior you are describing and there is no ambiguity. If that's true, then there is no reason for us to try to overengineer that method and potentially create ambiguity.

Reply:
No this factory is called even for the first time that zookeeper client needs to be created.. so we just need to create one and always return the same.

Reply:
Apart from that, we also need to be sure that no exception is raised in that method; otherwise we will encounter the same issue that makes the Controller to get blocked.